File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if(digits == "") return {};
        map<char,string> m;
        m['2'] = "abc";
        m['3'] = "def";
        m['4'] = "ghi";
        m['5'] = "jkl";
        m['6'] = "mno";
        m['7'] = "pqrs";
        m['8'] = "tuv";
        m['9'] = "wxyz";
        
        vector<string> ans;
        string temp = "";
        
        helper(0,temp,ans,m,digits);
        
        return ans;
    }
    
    void helper(int idx, string& temp,vector<string>& ans,map<char,string>& m,string& digits){
        if(idx == digits.size()){
            ans.push_back(temp);
            return;
        }
        
        for(auto i : m[digits[idx]]){
            temp.push_back(i);
            helper(idx+1,temp,ans,m,digits);
            temp.pop_back();
        }
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        ans = []
        d= {
            "2" : "abc",
            "3" :  "def",
            "4" : "ghi",
            "5" :  "jkl",
            "6" : "mno",
            "7" :  "pqrs",
            "8" : "tuv",
            "9" :  "wxyz"
        }

        arr =[d[num] for num in list(digits)]
        maxL = max(arr)
        
        def dfs(s,idx):
            if len(s) == len(digits):
                ans.append(str(s)) 
                return
            
            for i in range(0,len(arr[idx])):
                s += arr[idx][i]
                dfs(s,idx + 1)
                s = s[:-1]
        

        dfs("",0)
        return ans
